convert/safetensors.go

```diff
@@ -0,0 +1,283 @@
+package convert
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log/slog"
+	"os"
+	"regexp"
+	"slices"
+
+	"github.com/d4l3k/go-bfloat16"
+	"github.com/jmorganca/ollama/llm"
+	"github.com/pdevine/tensor"
+	"github.com/pdevine/tensor/native"
+	"github.com/x448/float16"
+)
+
+type SafetensorMetadata struct {
+	Type    string   `json:"type"`
+	Shape   []uint64 `json:"shape"`
+	Offsets []int    `json:"data_offsets"`
+}
+
+func ggmlTensorName(n string) (string, error) {
+	tMap := map[string]string{
+		"model.embed_tokens.weight":                           "token_embd.weight",
+		"model.layers.(\\d+).input_layernorm.weight":          "blk.$1.attn_norm.weight",
+		"model.layers.(\\d+).mlp.down_proj.weight":            "blk.$1.ffn_down.weight",
+		"model.layers.(\\d+).mlp.gate_proj.weight":            "blk.$1.ffn_gate.weight",
+		"model.layers.(\\d+).mlp.up_proj.weight":              "blk.$1.ffn_up.weight",
+		"model.layers.(\\d+).post_attention_layernorm.weight": "blk.$1.ffn_norm.weight",
+		"model.layers.(\\d+).self_attn.k_proj.weight":         "blk.$1.attn_k.weight",
+		"model.layers.(\\d+).self_attn.o_proj.weight":         "blk.$1.attn_output.weight",
+		"model.layers.(\\d+).self_attn.q_proj.weight":         "blk.$1.attn_q.weight",
+		"model.layers.(\\d+).self_attn.v_proj.weight":         "blk.$1.attn_v.weight",
+		"lm_head.weight":    "output.weight",
+		"model.norm.weight": "output_norm.weight",
+	}
+
+	v, ok := tMap[n]
+	if ok {
+		return v, nil
+	}
+
+	// quick hack to rename the layers to gguf format
+	for k, v := range tMap {
+		re := regexp.MustCompile(k)
+		newName := re.ReplaceAllString(n, v)
+		if newName != n {
+			return newName, nil
+		}
+	}
+
+	return "", fmt.Errorf("couldn't find a layer name for '%s'", n)
+}
+
+func ParseSafetensor(filename string, params *Params) ([]*llm.Tensor, error) {
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	var n uint64
+	if err := binary.Read(file, binary.LittleEndian, &n); err != nil {
+		return nil, err
+	}
+
+	slog.Info("ParseSafetensor", "size", n)
+
+	var b bytes.Buffer
+	if _, err := io.CopyN(&b, file, int64(n)); err != nil {
+		return nil, err
+	}
+
+	var headers map[string]SafetensorMetadata
+	if err := json.NewDecoder(&b).Decode(&headers); err != nil {
+		return nil, err
+	}
+
+	var keys []string
+	for k := range headers {
+		keys = append(keys, k)
+	}
+
+	slices.Sort(keys)
+
+	var tensors []*llm.Tensor
+	for _, key := range keys {
+		value := headers[key]
+
+		var kind uint32
+		switch len(value.Shape) {
+		case 0:
```

Comment:
This used to change kind to 0 when the shape value was 1. Is that no longer needed?